autopush/db.py

```diff
@@ -853,3 +861,130 @@ def clear_node(self, item):
             return True
         except ConditionalCheckFailedException:
             return False
+
+
+@attrs
+class DatabaseManager(object):
+    """Provides database access"""
+
+    storage = attrib()  # type: Storage
+    router = attrib()   # type: Router
+
+    metrics = attrib()  # type: IMetrics
+
+    message_tables = attrib(default=Factory(dict))  # type: Dict[str, Message]
+    current_msg_month = attrib(default=None)        # type: Optional[str]
+    current_month = attrib(default=None)            # type: Optional[int]
+
+    _message_prefix = attrib(default="message")  # type: str
+
+    @classmethod
+    def from_settings(cls, settings):
+        router_table = get_router_table(
+            settings.router_tablename,
+            settings.router_read_throughput,
+            settings.router_write_throughput
+        )
+        storage_table = get_storage_table(
+            settings.storage_tablename,
+            settings.storage_read_throughput,
+            settings.storage_write_throughput
+        )
+        get_rotating_message_table(
+            settings.message_tablename,
+            message_read_throughput=settings.message_read_throughput,
+            message_write_throughput=settings.message_write_throughput
+        )
+        metrics = autopush.metrics.from_settings(settings)
+        return cls(
+            storage=Storage(storage_table, metrics),
+            router=Router(router_table, metrics),
+            message_prefix=settings.message_tablename,
+            metrics=metrics
+        )
+
+    def setup(self, preflight_uaid):
+        # type: (str) -> None
+        """Setup metrics, message tables and perform preflight_check"""
+        # Used to determine whether a connection is out of date with current
+        # db objects. There are three noteworty cases:
+        # 1 "Last Month" the table requires a rollover.
+        # 2 "This Month" the most common case.
+        # 3 "Next Month" where the system will soon be rolling over, but with
+        #   timing, some nodes may roll over sooner. Ensuring the next month's
+        #   table is present before the switchover is the main reason for this,
+        #   just in case some nodes do switch sooner.
+        self.metrics.start()
+        self.create_initial_message_tables()
+        preflight_check(self.storage, self.router, preflight_uaid)
+
+    @property
+    def message(self):
+        """Property that access the current message table"""
+        return self.message_tables[self.current_msg_month]
+
+    @message.setter
+    def message(self, value):
+        """Setter to set the current message table"""
+        self.message_tables[self.current_msg_month] = value
+
+    def _tomorrow(self):
+        return datetime.date.today() + datetime.timedelta(days=1)
+
+    def create_initial_message_tables(self):
```

Comment:
forgot some of these and we do have some circular imports. this is going to start getting typing.TYPE_CHECKING ugly